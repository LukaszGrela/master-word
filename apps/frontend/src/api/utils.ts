import { TPartialGameState } from './types';

export const createGameState = (
  attempts: number,
  wordLength: number
): TPartialGameState => {
  const word = Array.from(Array(wordLength)).map(() => '');
  return Array.from(Array(attempts)).map(() => ({
    word: word.concat(),
  }));
};
