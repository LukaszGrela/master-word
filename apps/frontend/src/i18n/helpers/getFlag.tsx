import React from 'react';

export const getFlag = (language: string): React.ReactNode => {
  switch (language) {
    case 'pl':
      return <>ğŸ‡µğŸ‡±</>;
    case 'en':
      return <>ğŸ‡ºğŸ‡¸</>;
    default:
      return <>{language.toUpperCase()}</>;
  }
};
