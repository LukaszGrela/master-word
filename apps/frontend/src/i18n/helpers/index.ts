/* v8 ignore start */
export * from './loadTranslation';
/* v8 ignore stop */
