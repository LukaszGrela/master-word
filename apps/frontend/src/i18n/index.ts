/* v8 ignore start */
export * from './LanguageContext/hooks';
export { LanguageProvider } from './LanguageContext/index';
export type { ILanguageContext } from './LanguageContext/index';
/* v8 ignore stop */
