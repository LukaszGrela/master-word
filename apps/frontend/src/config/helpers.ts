import { AppStorage, EStorageKeys } from '@repo/utils';
import { IConfig } from './types';
import { ATTEMPTS, LANGUAGE, WORD_LENGTH } from '../game';

/**
 * Assure local storage is using enabled values from current config object
 * @param config Current Configuration
 */
export function updateLocalStorageConfig(config: IConfig) {
  console.log('config', config);
  const storage = AppStorage.getInstance();
  // game language
  const configLang =
    config.enabledLanguages.length > 0 ? config.enabledLanguages : [LANGUAGE];
  if (!storage.has(EStorageKeys.GAME_LANGUAGE)) {
    storage.setItem(EStorageKeys.GAME_LANGUAGE, configLang[0]);
  } else {
    // value exists, confirm it is allowed
    if (
      configLang.indexOf(storage.getItem(EStorageKeys.GAME_LANGUAGE)!) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_LANGUAGE, configLang[0]);
    }
  }

  // game word length
  const configLength =
    config.enabledLength.length > 0 ? config.enabledLength : [WORD_LENGTH];
  if (!storage.has(EStorageKeys.GAME_WORD_LENGTH)) {
    storage.setItem(EStorageKeys.GAME_WORD_LENGTH, `${configLength[0]}`);
  } else {
    // value exists, confirm it is allowed
    if (
      configLength.indexOf(
        Number(storage.getItem(EStorageKeys.GAME_WORD_LENGTH)!),
      ) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_WORD_LENGTH, `${configLength[0]}`);
    }
  }

  // game attempts
  const configAttempts =
    config.enabledAttempts.length > 0 ? config.enabledAttempts : [ATTEMPTS];
  if (!storage.has(EStorageKeys.GAME_ATTEMPTS)) {
    storage.setItem(EStorageKeys.GAME_ATTEMPTS, `${configAttempts[0]}`);
  } else {
    // value exists, confirm it is allowed
    if (
      configAttempts.indexOf(
        Number(storage.getItem(EStorageKeys.GAME_ATTEMPTS)!),
      ) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_ATTEMPTS, `${configAttempts[0]}`);
    }
  }

  // UI
  if (!storage.has(EStorageKeys.UI_LANGUAGE)) {
    storage.setItem(EStorageKeys.UI_LANGUAGE, LANGUAGE);
  }
}
