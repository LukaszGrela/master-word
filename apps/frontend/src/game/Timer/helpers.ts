import { toHMS } from '@repo/utils';

const prefix0 = (n: number): string => (n < 10 ? `0${n}` : `${n}`);

export const displayTime = (ms: number): string => {
  const { h, m, s } = toHMS(ms);

  if (!h && !m) {
    return `00:${prefix0(s >> 0)}`;
  }

  if (!h && m !== undefined) {
    return `${prefix0(m)}:${prefix0(s)}`;
  }

  if (h !== undefined && m !== undefined) {
    return `${prefix0(h)}:${prefix0(m)}:${prefix0(s)}`;
    /* v8 ignore start */
  }

  return '00:00'; // it should never reach here
};
/* v8 ignore stop */
