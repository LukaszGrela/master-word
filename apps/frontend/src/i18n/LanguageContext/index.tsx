import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { TSupportedLanguages } from '../../api';
import { IDictionary } from '../../types/common';
import { loadTranslation, replaceSubstituteMap } from '../helpers';
import { hasOwn } from '@repo/utils';
import { AppStorage } from '@repo/utils';
import { EStorageKeys } from '@repo/utils';

export interface ILanguageContext {
  loadedLanguage: TSupportedLanguages | undefined;

  loading: boolean;

  loadTranslation: (
    language: TSupportedLanguages,
    signal?: AbortSignal,
  ) => Promise<void>;

  getUIText: (
    textId: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    substitutes?: IDictionary<any>,
  ) => string;
}

export const LanguageContext = React.createContext<ILanguageContext>(null!);

export function LanguageProvider({ children }: { children: React.ReactNode }) {
  const [loading, setLoading] = useState(true);

  const [loadedLanguage, setLoadedLanguage] = useState<TSupportedLanguages>();
  const [translation, setTranslation] = useState<IDictionary<string>>({});

  const loadTranslationCtx = useCallback(
    async (language: TSupportedLanguages, signal?: AbortSignal) => {
      try {
        setLoading(true);
        const translations = await loadTranslation(language, signal);
        setLoadedLanguage(language);
        setTranslation(translations);
      } catch (error) {
        console.error(error);
      } finally {
        setLoading(false);
      }
    },
    [],
  );

  const getUIText = useCallback(
    (
      textId: string,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      substitutes?: IDictionary<any>,
    ): string => {
      if (hasOwn(translation, textId)) {
        const copy = translation[textId];
        if (substitutes) {
          return replaceSubstituteMap(copy, substitutes);
        }
        return copy;
      }

      return textId;
    },
    [translation],
  );

  const value = useMemo(
    () =>
      ({
        loading,
        loadedLanguage,
        getUIText,
        loadTranslation: loadTranslationCtx,
      }) as ILanguageContext,
    [getUIText, loadTranslationCtx, loadedLanguage, loading],
  );

  // initial load
  useEffect(() => {
    setLoading(true);
    const language =
      (AppStorage.getInstance().getItem(
        EStorageKeys.UI_LANGUAGE,
      ) as TSupportedLanguages) || 'pl';
    loadTranslation(language)
      .then((translations) => {
        setLoadedLanguage(language);
        setTranslation(translations);
      })
      .catch(console.error)
      .finally(() => {
        setLoading(false);
      });
  }, []);

  return (
    <LanguageContext.Provider value={value}>
      {children}
    </LanguageContext.Provider>
  );
}
