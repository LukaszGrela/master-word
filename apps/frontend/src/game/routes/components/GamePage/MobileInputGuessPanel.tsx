import { FC, useState, useEffect, useCallback, useMemo } from 'react';
import { useLanguage } from '../../../../i18n';
import { useGameContext } from '../../../context';
import { InputGuessPanel } from '../../../panel';
import { TClickAction } from '../../../panel/types';

export const MobileInputGuessPanel: FC = () => {
  const { loading, game, wordLength, error, guess } =
    useGameContext();
  const { getUIText: t } = useLanguage();

  const [showInputModal, setShowInputModal] = useState(false);
  useEffect(() => {
    const tapHandler = (e: MouseEvent) => {
      if ((e.target as HTMLElement).classList.contains('board')) {
        setShowInputModal(true);
      }
    };

    if (!loading) {
      document.addEventListener('click', tapHandler);
    }

    return () => {
      document.removeEventListener('click', tapHandler);
    };
  }, [loading]);

  const handlePanelAction = useCallback(
    (action: TClickAction, ...rest: unknown[]) => {
      if (action === 'guess') {
        // from input panel
        const [word] = rest as string[];

        setShowInputModal(false);
        guess(word);
      }
    },
    [guess],
  );

  const enrichedError = useMemo(() => {
    if (error && error.code === 6 /* INVALID_WORD */) {
      return {
        ...error,

        error: t('main-error-invalid-word'),
      };
    }
    return error;
  }, [error, t]);

  useEffect(() => {
    if (error && error.code === 6 /* INVALID_WORD */) {
      setShowInputModal(true);
    }
  }, [enrichedError, error]);

  const attempt = game?.attempt ?? 0;
  return (
    game && !game.finished && (
      <InputGuessPanel
        show={showInputModal}
        initWord={game.state[attempt].word.join('')}
        maxLength={wordLength}
        onClose={handlePanelAction}
        error={enrichedError}
      />
    )
  );
};
