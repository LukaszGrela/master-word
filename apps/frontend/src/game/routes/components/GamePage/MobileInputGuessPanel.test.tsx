import { describe, expect, it, vi } from 'vitest';
import { fireEvent } from '@testing-library/react';
import { IDictionary } from '@repo/common-types';
import { hasOwn } from '@repo/utils';
import { TGameRecord } from '@repo/backend-types';
import {
  contextRenderer,
  gameResponses,
} from '../../../context/__tests__/contextRenderer';
import { MockContext } from '../../../../i18n/LanguageContext/__tests__/contextRenderer';
import { MobileInputGuessPanel } from './MobileInputGuessPanel';
import { replaceSubstituteMap } from '../../../../i18n/helpers';

describe('game', () => {
  describe('routes', () => {
    describe('components', () => {
      describe('GamePage', () => {
        describe('MobileInputGuessPanel', () => {
          const translationPl: IDictionary<string> = {
            'continue-button': 'Kontynuuj grę',

            'input-modal-close-button': 'Ukryj',
            'input-modal-info-done': 'Naciśnij Gotowe by sprawdzić',
            'input-modal-info-start': 'Wpisz {maxLength} literowe słowo',
            'input-modal-label': 'Podaj słowo do odgadnięcia',
            'input-modal-ok-button': 'Gotowe',
            'main-desktop-info':
              'Zacznij wpisywać słowo, naciśnij Enter żeby potwierdzić.',
            'main-error-invalid-word': 'Niepoprawne słowo',
            'main-legend-correct': 'Dobra litera',
            'main-legend-incorrect': 'Zła litera',
            'main-legend-misplaced': 'Złe miejsce',
            'main-mobile-info':
              'Naciśnij linię z ikonką edycji by wprowadzić słowo.',
            'main-subtitle': 'Odgadnij słowo które mam na myśli...',
            'main-title': 'Mistrz Słowa',

            loading: 'Wczytuję...',
          };

          it('renders null when no game', () => {
            const { container } = contextRenderer(
              <MockContext
                context={{
                  loading: false,
                  getUIText: vi.fn().mockImplementation(
                    (
                      textId: string,
                      // eslint-disable-next-line @typescript-eslint/no-explicit-any
                      substitutes?: IDictionary<any> | undefined,
                    ): string => {
                      if (hasOwn(translationPl, textId)) {
                        const copy = translationPl[textId];
                        if (substitutes) {
                          return replaceSubstituteMap(copy, substitutes);
                        }
                        return copy;
                      }

                      return textId;
                    },
                  ),
                }}
              >
                <MobileInputGuessPanel />
              </MockContext>,
              {
                context: {
                  loading: false,
                },
              },
            );

            expect(container).toMatchSnapshot();
          });

          it('renders null when game finished', () => {
            const { container } = contextRenderer(
              <MockContext
                context={{
                  loading: false,
                  getUIText: vi.fn().mockImplementation(
                    (
                      textId: string,
                      // eslint-disable-next-line @typescript-eslint/no-explicit-any
                      substitutes?: IDictionary<any> | undefined,
                    ): string => {
                      if (hasOwn(translationPl, textId)) {
                        const copy = translationPl[textId];
                        if (substitutes) {
                          return replaceSubstituteMap(copy, substitutes);
                        }
                        return copy;
                      }

                      return textId;
                    },
                  ),
                }}
              >
                <MobileInputGuessPanel />
              </MockContext>,
              {
                context: {
                  loading: false,
                  game: gameResponses.gameWon.game as unknown as TGameRecord,
                },
              },
            );

            expect(container).toMatchSnapshot();
          });

          it('renders', () => {
            const { queryByText, getByTestId } = contextRenderer(
              <MockContext
                context={{
                  loading: false,
                  getUIText: vi.fn().mockImplementation(
                    (
                      textId: string,
                      // eslint-disable-next-line @typescript-eslint/no-explicit-any
                      substitutes?: IDictionary<any> | undefined,
                    ): string => {
                      if (hasOwn(translationPl, textId)) {
                        const copy = translationPl[textId];
                        if (substitutes) {
                          return replaceSubstituteMap(copy, substitutes);
                        }
                        return copy;
                      }

                      return textId;
                    },
                  ),
                }}
              >
                <div className="board" data-testid="board">
                  Click Me
                </div>
                <MobileInputGuessPanel />
              </MockContext>,
              {
                context: {
                  loading: false,
                  game: gameResponses.gameInit.game as unknown as TGameRecord,
                },
              },
            );

            let label = queryByText(translationPl['input-modal-label']);
            expect(label).toBeNull();

            const board = getByTestId('board');
            expect(board).toEqual(expect.anything());

            fireEvent.click(board);

            // should be opened
            label = queryByText(translationPl['input-modal-label']);
            expect(label).toEqual(expect.anything());
          });

          it('renders error', () => {
            const { queryByText, queryByDisplayValue } = contextRenderer(
              <MockContext
                context={{
                  loading: false,
                  getUIText: vi.fn().mockImplementation(
                    (
                      textId: string,
                      // eslint-disable-next-line @typescript-eslint/no-explicit-any
                      substitutes?: IDictionary<any> | undefined,
                    ): string => {
                      if (hasOwn(translationPl, textId)) {
                        const copy = translationPl[textId];
                        if (substitutes) {
                          return replaceSubstituteMap(copy, substitutes);
                        }
                        return copy;
                      }

                      return textId;
                    },
                  ),
                }}
              >
                <div className="board" data-testid="board">
                  Click Me
                </div>
                <MobileInputGuessPanel />
              </MockContext>,
              {
                context: {
                  loading: false,
                  game: gameResponses.gameNext.game as unknown as TGameRecord,

                  error: {
                    code: 6,
                    error: 'Error message',
                    details: 'CINDY',
                  },
                },
              },
            );

            let label = queryByText(translationPl['input-modal-label']);

            // should be opened
            label = queryByText(translationPl['input-modal-label']);
            expect(label).toEqual(expect.anything());
            expect(queryByDisplayValue('CINDY')).toEqual(expect.anything());
          });

          it('calls guess action', () => {
            const guess = vi.fn();

            const { queryByText, getByTestId } = contextRenderer(
              <MockContext
                context={{
                  loading: false,
                  getUIText: vi.fn().mockImplementation(
                    (
                      textId: string,
                      // eslint-disable-next-line @typescript-eslint/no-explicit-any
                      substitutes?: IDictionary<any> | undefined,
                    ): string => {
                      if (hasOwn(translationPl, textId)) {
                        const copy = translationPl[textId];
                        if (substitutes) {
                          return replaceSubstituteMap(copy, substitutes);
                        }
                        return copy;
                      }

                      return textId;
                    },
                  ),
                }}
              >
                <div className="board" data-testid="board">
                  Click Me
                </div>
                <MobileInputGuessPanel />
              </MockContext>,
              {
                context: {
                  loading: false,
                  game: gameResponses.gameNext.game as unknown as TGameRecord,
                  guess,
                },
              },
            );

            let label = queryByText(translationPl['input-modal-label']);
            expect(label).toBeNull();

            const board = getByTestId('board');
            expect(board).toEqual(expect.anything());

            fireEvent.click(board);

            // should be opened
            label = queryByText(translationPl['input-modal-label']);
            expect(label).toEqual(expect.anything());
            // type a word

            const input = getByTestId('input-modal-guess');
            // type word
            fireEvent.change(input, { target: { value: 'GUESS' } });

            // keyboard event
            input.focus();
            fireEvent.keyUp(document.activeElement || document.body, {
              key: 'Enter',
              code: 'Enter',
              charCode: 13,
            });

            expect(guess).toHaveBeenCalledTimes(1);
            expect(guess).toHaveBeenLastCalledWith('GUESS')
          });
        });
      });
    });
  });
});
