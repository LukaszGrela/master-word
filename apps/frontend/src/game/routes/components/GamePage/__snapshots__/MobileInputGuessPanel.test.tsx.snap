// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > routes > components > GamePage > MobileInputGuessPanel > renders null when game finished 1`] = `<div />`;

exports[`game > routes > components > GamePage > MobileInputGuessPanel > renders null when no game 1`] = `<div />`;
